### ISSUE TITLE
SPIKE: Determine manual process to be used for supporting wire payments

### ISSUE THREAD

**Issue by dallashuggins** [deptagency/algomart#139]
## Overview

There's going to be a requirement to offer customer service to those who need to pay using wire. What should an admin setup support?

## Additional context
Add a separate story for the implementation once this is complete. 

_Possible situations to handle:_
- Someone pays for the item but the wrong amount was sent
- Someone pays for the item but the bank denied it
- Someone wants to check at the status
- Revoking a claim to the pack
- Awarding the pack to another bidder
- Email notifications throughout process

**Comment by dallashuggins** [comment]
### Wire payments flow

- User creates bank account using checkout form and instructions are presented + emailed to them.
  - Expiration date is added to the payment
- Payment is initiated by the user
  - Scheduled task finds a matching pack payment
    - If a resolved status: Emails the application owner to let them know the status of the matching payment (i.e., paid, failed).
  - Scheduled task finds expired wire payments that are not in a resolved status (i.e., pending)
    - Checks for any matching payments. If resolved: Email the app owner with the status of the payment.
- Application owner logs into admin interface
- Application owner has a list of (expired?) payments
  - View email (and phone?) for user
    - Email/phone user in case something went wrong
    - Should this be an action they can take? Or just provide contact info for manual contact? 
  - Option to:
    - Review Circle API payments associated with the payment source ID. Actions include:
      - Revoke claim for associated pack
      - Awarding the pack to the user
      - Update expiration date for payment

### Admin options

#### Resolving wire payments
- Search through all wire payments in Circle's API (don't limit time; we use 24 hours in past for task)
- Find any matching wire payments in Circle's API; could require paging through.
- Show all wire payments associated with the bank account ID + statuses. 
  - No matching: Allow option to revoke claim to the pack or to update the expiration date for payment
  - Matching: Allow option to select the found payment as the payment for the pack
    - Removes expiration and updates the status of the payment and the external ID.
    - Alternatively, if the payment is found but still pending, they could update the expiration date.

#### Expired payments
- List all payments past the expiration date
  - If they have an external ID: they are linked to a payment in Circle's API.
    - Check the status
    - Option to extend expiration date if pending
    - Option to find new payments if there's a failed payment linked. 
      - For instance, if someone had successfully sent the funds but it took multiple tries. 
      - See _Resolving wire payments_ above
  - If there's no external ID:
    - See _Resolving wire payments_ above

#### All payments
- List all payments
  - Payments with a bank account ID is `wire`.
  - Payments with no bank account ID is `card` (may not have card association even if a card so cannot base off the payment card ID).
  - Sorted by date of creation (or update)
  - Display `status` in table
    - The list could include all unresolved statuses, or every payment that's been made.
- If a payment is selected (could be accordion table or individual pages), display more information + actions
  - Show associated pack details
  - Show associated user details
  - For wires: Option to search Circle API for any matching payments (i.e., see _Resolving wire payments_ above)
  - For cards: Option to revoke/send email about failed payment and possibly revoking
    - If something went wrong down the line and the payment goes from `confirmed` to `failed`, we could auto-revoke the packs, or we could display here and allows the application owner to revoke or send an email to try payment again.

#### All packs
- List all packs
  - Need ability to search by username and/or by the title/slug
  - Show if the pack is owned by someone
- Click on pack to see more details
  - For each edition of the pack:
    - Retrieve associated payment with the pack
    - List payment details (i.e., status, if associated with an external ID, if wire/card)
    - Options to revoke/send email/update expiration.
    - Option to check Circle's API more thoroughly (see _Resolving wire payments_ above) and associating a payment and awarding the pack

Listing all payments is more conducive to troubleshooting wires, but adding different options in case another route solved another use-case.

**Comment by dallashuggins** [comment]
Closing in place of #152

**Comment by dallashuggins** [comment]
### User stories for wire "unhappy" paths

_As a customer attempting to purchase a pack using wire transfers, I want to be able to change my bank account if purchasing using my other bank account fails_

- User is notified a wire payment was found with a failed status
- Customer support is notified of the failed payment status
- Customer support reaches out to user
  - Customer support finds the payment in the table
    - Search by user details (i.e., username) OR associated pack details (i.e., ID, slug, title)
  - User wants to provide new bank account
    - Completely manual process?
    - User provides bank account information to customer support
    - Action: Update bank account

_As a customer attempting to purchase a pack using wire transfers, I want to initiate a wire transfer after an unsuccessful wire transfer_

- User is notified a wire payment was found with a failed status
- Customer support is notified of the failed payment
- Customer support reaches out to user
  - Customer support finds the payment in the table
    - Search by user details (i.e., username) OR associated pack details (i.e., ID, slug, title)
  - User may want to try again
    - Customer support needs to reach out in the future to check status
      - Should we track anything? Set an expiration?
      - Likely manual process for now, where customer success periodically checks
    - Customer support finds the payment in the table
    - Customer support searches through matching payments
      - Action: Search for all matching Circle payments
      - IF payment is found:
       - Customer support finds a valid payment and updates the associated payment ID
         - Action: Update associated payment
     - IF payment not found or unsuccessful again:
       - Customer support does not find a valid payment and revokes the claim to the pack
         - Action: Revoke claim for pack

_As a customer attempting to purchase a pack using wire transfers, I want to initiate a wire transfer in the remaining amount after a successful wire transfer in the wrong amount_

- User is notified a wire payment was found for the incorrect amount
- Customer support is notified of the incorrect amount payment
- Customer support reaches out to user
  - Customer support finds the payment in the table
      - Search by user details (i.e., username) OR associated pack details (i.e., ID, slug, title)
      - Should we add a new status for payments for an incorrect amount?
  - User may want to refund the amount => manual process
  - User may want to send the remaining amount
    - Customer support needs to reach out in the future to check status
      - Should we monitor the time? Set an expiration? Likely manually initially.
    - Customer support finds the payment in the table
    - Customer support searches through matching payments
      - Action: Search for all matching Circle payments
    - IF payment is found:
      - Customer support finds a valid payment
        - Nothing needs to happen re: ownership since the pack owner is set, unless that changes to be claimed by, etc.
        - Update the status of the payment
          - Possibly add additional status for invalid amount, or mark as Failed at that point, and change to Paid
    - IF payment not found or unsuccessful again:
      - Customer support does not find a valid payment and revokes the claim to the pack
        - Action: Revoke claim for pack

### Actions

#### Search for all matching Circle payments
- Search Circle's API for all payments with a source ID matching the bank account ID associated with the payment
- List of payments in table format

#### Revoke claim
- Call endpoint to revoke the claim on the pack
  - Updates pack claimed by details
- Notify user they no longer own the pack

#### Update associated payment 
- Select a valid payment from the list of Circle payments
- Click a button to associate with the pack
  - Updates the external ID for the payment + status

#### Update bank account
- Update existing payment to remove pack association
- Create new bank account in Circle
- Create new wire payment in database, associated with this now unclaimed pack + user

**Comment by dallashuggins** [comment]
**Development**
- Endpoints
  - List all payments
    - Support for: pagination, sorting, filters
    - For all user stories
  - Search for all matching payments in Circle
    - Retrieves all payments in Circle matching the source ID, regardless of the price and status
    - Required to provide customer support the information they need in any situation. 
  - Revoke claim
    - Revoke a user’s claim to a pack
    - Notify the user
  - Update payment
    - Update the external ID and status for a payment
  - Update bank account
    - May not be a requirement, if changing bank accounts for the wire was a manual process
    - Remove pack association with existing payment, and add a new bank account and wire payment to associate with the pack.
    - For the As a customer attempting to purchase a pack using wire transfers, I want to be able to change my bank account if purchasing using my other bank account fails user story
- Tasks
  - Update payment statuses
    - For all payment types, send a notification to the user AND customer support when a resolved status is found (i.e., Failed or Paid).
  - Update bank account statuses
    - If the wrong amount is provided, update the status to Failed.
    - Should there be a different status for a wrong amount?
- Updates
  - New status for payments with wrong amount?
- Questions
  - Should we track expiration for the payment?
  - How should someone find a payment in the table?
